pyrefly/lib/tsp/server.rs

```diff
@@ -0,0 +1,140 @@
+/*
+ * Copyright (c) Meta Platforms, Inc. and affiliates.
+ *
+ * This source code is licensed under the MIT license found in the
+ * LICENSE file in the root directory of this source tree.
+ */
+
+use std::collections::HashSet;
+use std::sync::Arc;
+
+use dupe::Dupe;
+use lsp_server::Connection;
+use lsp_server::Request;
+use lsp_server::RequestId;
+use lsp_types::InitializeParams;
+use lsp_types::ServerCapabilities;
+
+use crate::commands::lsp::IndexingMode;
+use crate::lsp::queue::LspEvent;
+use crate::lsp::queue::LspQueue;
+use crate::lsp::server::LspServerFactory;
+use crate::lsp::server::LspServerInterface;
+use crate::lsp::server::ProcessEvent;
+use crate::lsp::server::capabilities;
+use crate::lsp::server::dispatch_lsp_events;
+use crate::lsp::transaction_manager::TransactionManager;
+
+/// TSP server that delegates to LSP server infrastructure while handling only TSP requests
+pub struct TspServer {
+    pub inner: Box<dyn LspServerInterface>,
+}
```

Comment:
I think you can do this to avoid the heap allocation:
```suggestion
pub struct TspServer<T: LspServerInterface> {
    pub inner: T,
}
```